### ISSUE TITLE
Add a do not compile Modeline

### ISSUE THREAD

**Issue by 927589452** [lervag/vimtex#1440]
**Is your feature request related to a problem? Please describe it.**
I work with splitted files, but a few are shared between projects, so a `%! TEX root = ./file.tex` doesn't work. When in split mode `vimtex` tries to compile these, if they are rexognized as tex and as the aren't valid documents it generates an error.

The `vim` session

####################

---------------------------------------
```tex
% file: reused.tex

Something to be reused
```
---------------------------------------
```tex
% file: main.tex
\documentclass[draft,debug]{article}
\input{reused.tex}
\begin{document}
some more stuff
\end{document}
```
---------------------------------------
```
Vimtex Quickfix
```
---------------------------------------

####################

With continuus compile running

**Describe the solution you'd like**
add a flag like `%!TEX NOCOMPILE` that prevents vimtex from compiling this document

**Describe alternatives you've considered**
adding one of the compiling files as `tex root`, but this is rather suboptimal

**Comment by lervag** [comment]
I'm sorry, but I don't understand your issue:

> Add a do not compile Modeline

IMHO, we should distinguish the TeX directives (e.g. `%! TeX program`) from Vim modelines (e.g. `% vim: filetype=tex`). The latter is a general Vim feature, while the former is LaTeX specific and recognized by multiple editors.

> I work with splitted files, ...

What do you mean? Do you mean projects that consist of multiple files? I assume that in the following.

> ... but a few are shared between projects, so a %! TEX root = ./file.tex doesn't work. When in split mode vimtex tries to compile these, if they are rexognized as tex ...

Ok:
* What do you mean by "split mode"?
* You can add `let g:tex_flavor = 'latex'` to your vimrc file to make all `.tex` files be recognized as LaTeX instead of plainTeX. Then you also don't need the modelines for setting the filetype.

> ... and as the aren't valid documents it generates an error.

I think the sentence must be wrong, because it does not make sense to me. "as the aren't valid"? What is not valid?

> Relevant files:

Your list of files is confusing. Some comments:
- You should always give file listings a proper filename to distinguish them and so that we may refer to them.
- The file that is reused is not included anywhere...
- The third listing does not make sense. What is the point of it?

> With continuous compile running

???

>  add a flag like `%!TEX NOCOMPILE` that prevents vimtex from compiling this document.

What does such a flag mean, though? How should it work? In any case, I think we should hold off this discussion until I completely understand your main issue.

**Comment by 927589452** [comment]
> I'm sorry, but I don't understand your issue:
> 
> > Add a do not compile Modeline
> 
> IMHO, we should distinguish the TeX directives (e.g. `%! TeX program`) from Vim modelines (e.g. `% vim: filetype=tex`). The latter is a general Vim feature, while the former is LaTeX specific and recognized by multiple editors.

Yes, sorry I mixed them up; was thinking about a TeX directive.
> 
> > I work with splitted files, ...
> 
> What do you mean? Do you mean projects that consist of multiple files? I assume that in the following.

Yes a project consisting of multiple files.
> 
> > ... but a few are shared between projects, so a %! TEX root = ./file.tex doesn't work. When in split mode vimtex tries to compile these, if they are rexognized as tex ...
> 
> Ok:
> 
>     * What do you mean by "split mode"?
> 

the internal `:split` mode of vim; I have 3 buffers, one with a valid document, one with an "invalid" document (as there is no `\begin{document}`, and the quickfix buffer showing the LaTeX errors

>     * You can add `let g:tex_flavor = 'latex'` to your vimrc file to make all `.tex` files be recognized as LaTeX instead of plainTeX. Then you also don't need the modelines for setting the filetype.
>
This makes sense and safes me from using the modelines; thanks
> > ... and as the aren't valid documents it generates an error.
> 
> I think the sentence must be wrong, because it does not make sense to me. "as the aren't valid"? What is not valid?

It confuses me the morning after too; the problem is, vimtex tried to compile a document, which is not compileable and overwrote the errors I wanted to see from the other document.

> > Relevant files:
> 
> Your list of files is confusing. Some comments:
> 
>     * You should always give file listings a proper filename to distinguish them and so that we may refer to them.
> 
>     * The file that is reused is not included anywhere...
Will fix this after sending the answer, though the latter is kind of the point.

>     * The third listing does not make sense. What is the point of it?
> 
Here we got the error; I was meant to replace a real screenshot to show, which 3 buffers were running

> > With continuous compile running
> 
> ???

`vimtex: Compiler is running`
Basically vimtex will try to compile every tex file

> 
> > add a flag like `%!TEX NOCOMPILE` that prevents vimtex from compiling this document.
> 
> What does such a flag mean, though? How should it work? I
This file should not be compiled (on its own) by vimtex and it should be ignored for the quickfix buffer; I tried to reproduce the problem and found it beeing connected to the habit of using `\le` twice to reload the quickfix buffer, which means, if I am in the input file the quickfix errorlog is pointing to this files error log.

**Comment by 927589452** [comment]
so basically there are multiple files using the `reused.tex`, but I know vimtex won't find the connection for each of them, because there are tweaks in the `\input@path`, so I want to give vimtex a hint not to compile this file, but use everything except the compilation dependend tools for this file and I would like vimtex to remember which was the last "mainfile"  to provide the quickfiy

(writing this down it seems to be rather big change, so feel free to say no)
(on a second thought, this could be done, inside the main finding method; if there is no main defined and this "flag" is set for the file, reuse the last main file)

**Comment by lervag** [comment]
> the internal :split mode of vim; I have 3 buffers, one with a valid document, one with an "invalid" document (as there is no \begin{document}, and the quickfix buffer showing the LaTeX errors

Ok. FYI: The proper name for this (wrt. Vim) is having three windows, not "split mode".

> This makes sense and safes me from using the modelines; thanks

Good, then I'll remove those from your original post.

> It confuses me the morning after too; the problem is, vimtex tried to compile a document, ...

No, you try to compile, not vimtex. That is, it is you who issue the compilation command. Vimtex just tries to do what is expected. This is the main reason that I want to understand your complete setup.

> Basically vimtex will try to compile every tex file

No, vimtex will try to compile the current project. When you open a LaTeX file, vimtex parses the document and looks for the main file if the current file is not the main file. A main file is the one that contains `\documentclass`.

Thus, your issue is likely with detecting the main file.

> so basically there are multiple files using the reused.tex, but I know vimtex won't find the connection for each of them, because there are tweaks in the \input@path, so I want to give vimtex a hint not to compile this file, but use everything except the compilation dependend tools for this file and I would like vimtex to remember which was the last "mainfile" to provide the quickfiy

----

The only real thing missing here is a proper minimal example. I would expect the following is a good example. First, given a minimal vimrc file and three LaTeX files:

```vim
set nocompatible
let &rtp = '~/.vim/bundle/vimtex,' . &rtp
let &rtp .= ',~/.vim/bundle/vimtex/after'
filetype plugin indent on
syntax enable

let g:tex_flavor = 'latex'
```

```tex
% file: included.tex
%! TeX root = NONE

Something to be included
```

```tex
% file: main1.tex
\documentclass[draft,debug]{article}
\input{included.tex}
\begin{document}
some more stuff
\end{document}
```

```tex
% file: main2.tex
\documentclass[draft,debug]{article}
\input{included.tex}
\begin{document}
some more stuff
\end{document}
```

The desired feature is to allow something like `TeX root = NONE`, which should have the following behaviour:

1. Open `main1.tex`, then open `included.tex`. `included.tex` should now be considered part of the `main1.tex` project.

2. Open `main2.tex`, then open `included.tex`. `included.tex` should now be considered part of the `main2.tex` project.

3. Open `included.tex` in a fresh session. `included.tex` should be its own main file, that is, there should be no automatic detection.

Do you agree with my description?

**Comment by lervag** [comment]
Just to add: The current (automatic) behaviour is to search for the main file and select the first candidate. E.g., in my suggested example, `main1.tex` is detected.

One possible improvement to automatic detection is to look at the list of open projects. That is, when we open `included.tex`, it will detect two possible candidates. Instead of choosing the first, we could look at the vimtex data. If e.g. `main2.tex` was already open, then use this as the main.

A sort of "drawback" is that this method will always find a main file (if available), and the one that is used may not be the one that one wants if the choice becomes ambiguous. E.g. if no main file is already open, or if both are open.

**Comment by 927589452** [comment]
> The only real thing missing here is a proper minimal example. I would expect the following is a good example. First, given a minimal vimrc file and three LaTeX files:
>
> The desired feature is to allow something like `TeX root = NONE`, which should have the following behaviour:
>
> 1. Open `main1.tex`, then open `included.tex`. `included.tex` should now be considered part of the `main1.tex` project.
>
> 2. Open `main2.tex`, then open `included.tex`. `included.tex` should now be considered part of the `main2.tex` project.
>
Yes, thanks for extracting my idea from my incoherent rumblings.
> 3. Open `included.tex` in a fresh session. `included.tex` should be its own main file, that is, there should be no automatic detection.
>
I am not sure about this; maybe even issue a warning, when starting the compiler/opening a fresh session.

**Comment by 927589452** [comment]
> Just to add: The current (automatic) behaviour is to search for the main file and select the first candidate. E.g., in my suggested example, `main1.tex` is detected.

I just realized for my usecase I could probably get away with a regexp for finding the main file (I use a `git submodule` to bring reused components to all my tex files)

> One possible improvement to automatic detection is to look at the list of open projects. That is, when we open `included.tex`, it will detect two possible candidates. Instead of choosing the first, we could look at the vimtex data. If e.g. `main2.tex` was already open, then use this as the main.

This sounds good to me.
> A sort of "drawback" is that this method will always find a main file (if available), and the one that is used may not be the one that one wants if the choice becomes ambiguous. E.g. if no main file is already open, or if both are open.

Maybe use a variable, which is set, when ever a file with a legitimate tex root is added?

* Open `main1.tex`, then open `included.tex`. `included.tex` should now be considered part of the `main1.tex` project. Now open `main2.tex` from the `main1.tex`-window and switch to the `included.tex`-window: `included.tex` is now considered to be part of `main2.tex`.
 
* Open ` main1.tex`, open `main2.tex`, then open `included.tex`. `included.tex` should now be considered part of the `main2.tex` project.
 
* Open `included.tex` in a fresh session. `included.tex` should warn about there not being any `main.tex` file. Afterall we would declare this file to be incompilable; maybe it would be even better to issue this warning by the `VimtexCompile` call, as it is not relevant unless we want to compile

**Comment by lervag** [comment]
> - Open main1.tex, then open included.tex. included.tex should now be considered part of the main1.tex project. Now open main2.tex from the main1.tex-window and switch to the included.tex-window: included.tex is now considered to be part of main2.tex.

I think changing the "owner" of `included.tex` is too hard to implement compared to the added benefit. So I won't do it. Sorry.

> - Open main1.tex, open main2.tex, then open included.tex. included.tex should now be considered part of the main2.tex project.

Yes, this is OK.

> - Open included.tex in a fresh session. included.tex should warn about there not being any main.tex file. Afterall we would declare this file to be incompilable; maybe it would be even better to issue this warning by the VimtexCompile call, as it is not relevant unless we want to compile

Compilation is not really relevant here, because this issue is about how vimtex detects the proper main file for a given included file. It impacts many more features than just compilation. As such, let's just keep compilation out of it.

----

Based on our discussion so far, here is my proposal for a refined behaviour when loading `included.tex` which is included in both `main1.tex` and `main2.tex`:

- Open `included.tex`. `included.tex` will detect two possible candidates in `main1.tex` and `main2.tex` and ask the user to choose which one to use (if any).

- Open `main1.tex`, then `included.tex`. `included.tex` will detect `main1.tex` as the main file.

- Open `main2.tex`, then `included.tex`. `included.tex` will detect `main2.tex` as the main file.

- Open `main1.tex`, then `main2.tex`, then `included.tex`. `included.tex` will detect `main2.tex` as the main file.

- Open `main2.tex`, then `main1.tex`, then `included.tex`. `included.tex` will detect `main1.tex` as the main file.

Note: In my current proposal, it is not possible to change which project `included.tex` will belong to after it has fully loaded. I don't think this should be so important, since you can just quit Vim and open again if you need to edit `included.tex` as part of the other main file. Still, if you find that it would be very convenient to change the main file after initialization, then I propose that you open a new issue for this (please refer here), _after_ this issue has been resolved.

**Comment by 927589452** [comment]
this sounds good overall

**Comment by lervag** [comment]
Done. Let me know what you think!

**Comment by sga001** [comment]
Hello, commit b933bf2 suddenly broke things for me.

Here's the issue: when I open a particular file in my project with `vim --servername VIM -u minivimrc filename.tex`,  vim never finishes loading. It just hangs forever. I did a binary search to find out which commit did it, and it was the above one.

I'm happy to provide more information, I just don't know how to get any useful info since vim freezes and I cannot call VimtexCompileOutput, or VimtexInfo. 

My main file's name is: main.tex

Some observations:

1) if I open any other file in the same project, vim opens just fine. Then, if I do `:e filename.tex`, the trouble file (`filename.tex`) opens fine.

2) there is another file in the project with a similar name `filename-something.tex`. If I rename that file to `somethingelse.tex`, then suddenly `filename.tex` opens just fine. This suggest that maybe there is some bug with name parsing that causes some kind of infinite loop?

**Comment by lervag** [comment]
That's interesting. Would you be able to create a minimal LaTeX example for me to test on? If I can reproduce it, I should be able to fix it fast.

**Comment by sga001** [comment]
Yes, the following causes the issue. You'll need 4 files. 

**main.tex:**
  ```
\documentclass[10pt,twocolumn]{article}
\begin{document}
\input{broken}
\input{a}
\end{document}
```

**broken.tex**: can contain anything.

**a.tex**:  ``\input{broken-notreally}``

**broken-notreally.tex**: can contain anything.

If you try to open broken.tex, it'll hang forever.

**Comment by lervag** [comment]
Hmm. For me, it is not broken. It works with `main.tex`, `broken.tex` and `a.tex`, but for `broken-notreally.tex` there will be a question about which file to open.

**Comment by sga001** [comment]
Ah okay. I think the issue is with "autocmd BufNewFile,BufRead *.tex". Try this min vimrc:

```
  set nocompatible
  let &runtimepath  = '~/.vim/bundle/vimtex,' . &runtimepath
  let &runtimepath .= ',~/.vim/bundle/vimtex/after'
  filetype plugin indent on
  syntax enable

  autocmd BufNewFile,BufRead *.tex set ft=tex
  autocmd FileType tex set spell
```

**Comment by lervag** [comment]
I'm sorry, I did something wrong first, I was able to reproduce the problem and I think it should be fixed now.
